pkg/providers/instanceprofile/instanceprofile.go

```diff
@@ -0,0 +1,159 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instanceprofile
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/iam/iamiface"
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/apis/settings"
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	awserrors "github.com/aws/karpenter/pkg/errors"
+	"github.com/aws/karpenter/pkg/utils"
+)
+
+var (
+	instanceStateFilter = &ec2.Filter{
+		Name:   aws.String("instance-state-name"),
+		Values: aws.StringSlice([]string{ec2.InstanceStateNamePending, ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping, ec2.InstanceStateNameStopped, ec2.InstanceStateNameShuttingDown}),
+	}
+)
+
+type Provider struct {
+	region string
+	iamapi iamiface.IAMAPI
+	ec2api ec2iface.EC2API
+	cache  *cache.Cache
+}
+
+func NewProvider(region string, iamapi iamiface.IAMAPI, ec2api ec2iface.EC2API, cache *cache.Cache) *Provider {
+	return &Provider{
+		region: region,
+		iamapi: iamapi,
+		ec2api: ec2api,
+		cache:  cache,
+	}
+}
+
+func (p *Provider) Create(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (string, error) {
+	tags := lo.Assign(nodeClass.Spec.Tags, map[string]string{
+		fmt.Sprintf("kubernetes.io/cluster/%s", settings.FromContext(ctx).ClusterName): "owned",
+		corev1beta1.ManagedByAnnotationKey:                                             settings.FromContext(ctx).ClusterName,
+		v1beta1.LabelNodeClass:                                                         nodeClass.Name, v1.LabelTopologyRegion: p.region,
+	})
```

Comment:
musing: It would be nice if we could have a nodeclassutils.GetKarpenterTags() that would return this for us so we don't have to do it each time in multiple places in the code. Would be nice to align this elsewhere too, like in instance.go and launchtemplate.go

Reply:
Agreed. I could see this coming in a follow-on.